### PR TITLE
Spectator mode

### PR DESCRIPTION
**Pull request by FoxLoveFire**

### CODE REVIEW

**Comment by Nanowolf4** [comment]
Alright, thank you for your pull request. Let's take a closer look at it...

**Comment by Nanowolf4** [comment]
In the init.lua file, you specified the GPL-3.0-or-later, but the LICENSE.txt contains the MIT license.

File: mods/other/spectator/init.lua [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    give_to_admin = false,
+    give_to_singleplayer = false
+})
+
+-- Enabled spectator mode
+spectator.on = function (player)
+    local name = player:get_player_name()
+    local meta = player:get_meta()
+
+    meta:set_string(name, minetest.privs_to_string(minetest.get_player_privs(name), ","))
```
Comment (Nanowolf4) [inline mods/other/spectator/init.lua:25]: Makes sense to store and modify only necessary privileges like fly, shout, noclip, fast 

Keep in mind that while a player is in spectator mode, players privileges cannot be changed in the normal way.
For example, you had the ban privilege, enter spectator mode, admin revoke your ban priv, you left spectator mode, the ban privilege is still with you

File: mods/other/spectator/init.lua [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        visual = "",
+        show_on_minimap = false,
+        pointable = false,
+    })
+    player:set_nametag_attributes({color={a=0},text = " "})
+    player:set_nametag_attributes{text = "\0"}
+    player:set_armor_groups({immortal = 1})
+    player:get_inventory():set_list("main", {})
+    player:get_inventory():set_list("craft", {})
+    player:get_inventory():set_list("craftpreview", {})
```
Comment (Nanowolf4) [inline mods/other/spectator/init.lua:44]: The inventory doesn't have minecraft-style crafttable, no needed.

File: mods/other/spectator/init.lua [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+end)
+
+ctf_api.register_on_match_start(function()
+    for _, v in pairs(minetest.get_connected_players()) do
+        spectator.formspec(v:get_player_name())
+    end
+end)
+
+spectator.formspec = function (playername)
+    local formspec = "size[8,3]bgcolor[#080808BB;true]" .. default.gui_bg .. default.gui_bg_img .. [[
```
Comment (Nanowolf4) [inline mods/other/spectator/init.lua:82]: A good practice is to specify the formspec version by adding the element `formspec_version[<version>]` and using actual versions of it

File: mods/other/spectator/init.lua [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        button_exit[3.5,2.3;2,0.8;no;No]
+        button_exit[5.5,2.3;2,0.8;cancel;Cancel]
+        ]]
+    formspec = formspec .. "label[2.3,0.7;" .. "Watch the game?" .. "]"
+
+    minetest.after(0.2, minetest.show_formspec, playername, "Watch the game?", formspec)
+    
+end
+
+minetest.register_on_player_receive_fields(function(player, formname, fields)
```
Comment (Nanowolf4) [inline mods/other/spectator/init.lua:95]: - There's is no formname check.
- Return true it is handled.

File: mods/other/spectator/init.lua [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        pointable = true,
+    })
+    player:set_nametag_attributes {text = name}
+    player:set_armor_groups({immortal = 0})
+
+    spectator.in_[name] = nil
+end
+
+ctf_api.register_on_match_end(function()
+    for i,_ in pairs(spectator.in_) do
```
Comment (Nanowolf4) [inline mods/other/spectator/init.lua:70]: Recommended to name the variable i as name or player_name for clarity.

File: mods/other/spectator/init.lua [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        spectator.off(minetest.get_player_by_name(i))
+    end
+end)
+
+ctf_api.register_on_match_start(function()
+    for _, v in pairs(minetest.get_connected_players()) do
```
Comment (Nanowolf4) [inline mods/other/spectator/init.lua:76]: The same recommendation as above

File: mods/other/spectator/init.lua [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        spectator.formspec(v:get_player_name())
+    end
+end)
+
+spectator.formspec = function (playername)
+    local formspec = "size[8,3]bgcolor[#080808BB;true]" .. default.gui_bg .. default.gui_bg_img .. [[
+        hypertext[2.3,0.1;5,1;title;<b>Spectator mode<\b>]
+        image[0,0;2,2;question.png]
```
Comment (Nanowolf4) [inline mods/other/spectator/init.lua:84]: The image name "question.png" should include the mod name first

File: mods/other/spectator/init.lua [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+})
+
+minetest.register_on_joinplayer(function(player)
+    local name = player:get_player_name()
+    local meta = player:get_meta()
+    if meta:get_int("spectator") == 1 then
+        minetest.set_player_privs(name, minetest.string_to_privs(meta:get_string(name), ","))
+		meta:set_int("spectator", 0)
+    end 
+    spectator.formspec(name)
```
Comment (Nanowolf4) [inline mods/other/spectator/init.lua:121]: Putting this formspec upon joining the game will be annoying. Should it be activated by a chat command instead?

**Comment by Nanowolf4** [comment]
In case of a server crash, the player will be with **spectators privileges**

**Comment by Nanowolf4** [comment]
Despite of active spectator mode, the player is still a member of the team. Take the player out of the match

File: mods/other/spectator/textures/question.png
Comment (Nanowolf4) [inline mods/other/spectator/textures/question.png:1]: The image does not match the style of the game. Recommended to follow the icon style as game appearance

File: mods/other/spectator/init.lua [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    local formspec = "size[8,3]bgcolor[#080808BB;true]" .. default.gui_bg .. default.gui_bg_img .. [[
+        hypertext[2.3,0.1;5,1;title;<b>Spectator mode<\b>]
+        image[0,0;2,2;question.png]
+        button_exit[1.5,2.3;2,0.8;yes;Yes]
+        button_exit[3.5,2.3;2,0.8;no;No]
+        button_exit[5.5,2.3;2,0.8;cancel;Cancel]
+        ]]
+    formspec = formspec .. "label[2.3,0.7;" .. "Watch the game?" .. "]"
+
+    minetest.after(0.2, minetest.show_formspec, playername, "Watch the game?", formspec)
```
Comment (Nanowolf4) [inline mods/other/spectator/init.lua:91]: Why is there `minetest.after`...?

File: mods/other/spectator/textures/question.png
Comment (Nanowolf4) [inline mods/other/spectator/textures/question.png:1]: The texture name should be like that "spectator_question.png"

File: mods/other/spectator/init.lua [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        button_exit[1.5,2.3;2,0.8;yes;Yes]
+        button_exit[3.5,2.3;2,0.8;no;No]
+        button_exit[5.5,2.3;2,0.8;cancel;Cancel]
+    ]]
+    formspec = formspec .. "label[2.3,0.7;" .. "Watch the game?" .. "]"
+
+    minetest.show_formspec(playername, "f_spec", formspec)
+end
+
+minetest.register_on_player_receive_fields(function(player, formname, fields)
```
Comment (Nanowolf4) [inline mods/other/spectator/init.lua:84]: Should return true if formname is what you need to handle

File: mods/other/spectator/init.lua [hunk 11]
```diff
@@ -0,0 +1,112 @@
+-- SPDX-License-Identifier: GPL-3.0-or-later
+-- Copyright (c) 2024 Ivan Shkatov (Maintainer_) ivanskatov672@gmail.com
+spectator = {}
+spectator.spectators = {}
+spectator.texture = "question.png"
```
[the remark this answers is not in the document]
Comment (Nanowolf4) [reply mods/other/spectator/init.lua:5]: spectator.texture is unused.

File: mods/other/spectator/init.lua (hunk 10 above)
Comment (Nanowolf4) [reply mods/other/spectator/init.lua:85]: The same recommendation as above
https://github.com/Minetest-JMA-group/jma-capturetheflag/pull/10#discussion_r1743733961

File: mods/other/spectator/init.lua [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+ctf_api.register_on_match_end(function()
+    for player_name,_ in pairs(spectator.spectators) do
+        spectator.off(minetest.get_player_by_name(player_name))
+    end
+end)
+
+ctf_api.register_on_match_start(function()
+    for _, player in pairs(minetest.get_connected_players()) do
+        spectator.formspec(player:get_player_name())
```
[the remark this answers is not in the document]
Comment (FoxLoveFire) [reply mods/other/spectator/init.lua:66]: No. Only formspec.

File: mods/other/spectator/init.lua [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    end
+end)
+
+spectator.formspec = function (playername)
```
Comment (Nanowolf4) [inline mods/other/spectator/init.lua:70]: The function declaration must have this syntax:
function spectator.formspec(playername)

**Review by Nanowolf4 — COMMENTED** [review]
As mentioned earlier, a player in spectator mode shouldn't be a member of a team. It is necessary to remove the player from the current game

**Comment by fancyfinn9** [comment]
I am still concerned about cheating possibilities, where a dedicated teammate could view the match in spectator mode so their team could know where everything and the other team is. Is there a way to avoid this?

File: mods/other/spectator/init.lua [hunk 14]
```diff
@@ -2,7 +2,7 @@
 -- Copyright (c) 2024 Ivan Shkatov (Maintainer_) ivanskatov672@gmail.com
 spectator = {}
 spectator.spectators = {}
-spectator.build_time = true
+local spectator.build_time = true
```
Comment (Nanowolf4) [inline mods/other/spectator/init.lua:5]: Incorrect syntax.
Should be ` local build_time = true`

**Comment by FoxLoveFire** [comment]
> I am still concerned about cheating possibilities, where a dedicated teammate could view the match in spectator mode so their team could know where everything and the other team is. Is there a way to avoid this?

How?

**Pull request closed without merging**